### ISSUE TITLE
Blockwise with `concatenate=True` broken with LocalCluster + `optimize_graph=False` 

### ISSUE THREAD

**Issue by JSKenyon** [dask/dask#7449]
**What happened**:
When using a LocalCluster (and probably the distributed scheduler in general), setting `optimize_graph=False` causes strange (and incorrect) behaviour in Blockwise with `concatentate=True`. 

**What you expected to happen**:
Blockwise with `concatenate=True` should function as normal.

**Minimal Complete Verifiable Example**:

```python
import dask.array as da
from distributed import LocalCluster, Client


def f(x, y):

    print(y)

    return x


if __name__ == '__main__':

    cluster = LocalCluster(processes=False,
                           n_workers=1,
                           threads_per_worker=1)

    client = Client(cluster)

    x = da.array([0])
    y = da.array([[0]])

    z = da.blockwise(
        f, ("i"),
        x, ("i"),
        y, ("ij"),
        dtype=x.dtype,
        concatenate=True)

    da.compute(z, optimize_graph=False)
```

**Anything else we need to know?**:
The example above will print something like:
```
(<function concatenate_axes at 0x7f153d45c310>, [array([[0]])], [1])
```
This is not what is expected - we expect an array. Note that the above example will work as normal when:
- `optimize_graph=True`
- A different scheduler is used.

Apologies if this should have been posted on `distributed` instead.

**Environment**:

- Dask version: 2021.3.0
- Distributed version: 2021.3.0
- Python version: 3.8.8
- Operating System: Pop!_OS 18.04 LTS
- Install method (conda, pip, source): pip

**Comment by mrocklin** [comment]
cc @rjzamora

**Comment by rjzamora** [comment]
Thanks for raising @JSKenyon - I'll check it out

**Comment by rjzamora** [comment]
Hmm - This issues is exposing a general HLG/Blockwise challenge related to complex/nested (i.e. inlined) task serialization...

Since the primary callable object is already serialized by the time it reaches the scheduler, in `make_blockwise_graph`, we are replacing the usual `dumps_task` call with [an explicit `{"function": , "args": }` dictionary](https://github.com/dask/dask/blob/f6e9e81903bd2416125a97cc3856c63bb792bddf/dask/blockwise.py#L927).   In cases without task inlining, this is fine, because it is exactly what [dumps_task](https://github.com/dask/distributed/blob/98148cbfee818889f683c4be0e591d33a59ffdf8/distributed/worker.py#L3515) would do anyway.  When there **is** a "complex" task, however, `dumps_task` uses `to_serialize` instead (and the Blockwise behavior is just wrong).

@jrbourbeau @jakirkham - I'll look into this, but do you have any high-level intuition? Might it be possible to use `to_serialize` in this case, even if the primary function is already serialized?

**Comment by JSKenyon** [comment]
@rjzamora @mrocklin I think this might be back in the latest release. My original reproducer is breaks again.

**Comment by JSKenyon** [comment]
I'll amend that - the reproducer breaks, but in a different way. Should I open a new issue?

**Comment by rjzamora** [comment]
Thanks for reporting @JSKenyon !  I will try to reproduce soon.  However, since your example is using `processes=False`, you may be running into the same problem as #7777 - Note that Mads' PR in distributed will hopefully address this once it is merged and used in Dask (see #7815).

**Comment by JSKenyon** [comment]
Hi @rjzamora! I know it has been a while but this reproducer is still broken (and makes debugging some of my code very difficult). Should I open a new issue?

**Comment by rjzamora** [comment]
Thanks for following up on this @JSKenyon - Rather than re-opening this issue, I submitted #8581 to focus a bit more on the root serialization problem.
